enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,580 @@
+---
+title: operator-status
+authors:
+  - "@awgreene"
+reviewers:
+  - "@ecordell"
+approvers:
+  - "@kevinrizza"
+creation-date: 2020-05-19
+last-updated: 2020-06-10
+status: implementable
+see-also:
+  - "N/A"  
+replaces:
+  - "N/A"
+superseded-by:
+  - "N/A"
+---
+
+# operator-status
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+The goal of this enhancement is to allow operators managed by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) operator to communicate conditions and states that OLM would otherwise be unable to infer. The presence of certain conditions should influence OLM behavior to protect the cluster.
+
+## Motivation
+
+OLM currently infers the state of an operator by inspecting the Kubernetes resources that define the operator. OLM does not provide operators with the means to communicate complex states that cannot be inferred without operator specific knowledge.
+
+Operators managed by OLM would benefit from the ability to communicate their state directly to OLM through a supported channel. Operators could communicate complex conditions and influence how OLM manages the operator, dramatically improving user experience.
+
+A core requirement of this enhancement is to provide operators with the ability to communicate "Upgrade Readiness", providing operators with the ability to signal to OLM that they should not be upgraded at this time. Consider the following scenario based on OLM today:
+
+>An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
```

Comment:
This example conflicts with the:

> OLM cannot guarantee that some outside force will not interrupt the services of the operator and will not provide the operator with the ability to communicate that it should not be "Interrupted".

Non-goal.  Personally, I prefer the non-goal.  The solution to "things break when I do dangerous, irrecoverable things and get updated part-way through" is "don't do dangerous, irrecoverable things".  A better example for this enhancement might be "later versions of my operator, which you might be considering updating to, are not compatible with the current cluster state.  If you update me to those versions, I will break".  Or something like that?

Reply:
>ersonally, I prefer the non-goal. The solution to "things break when I do dangerous, irrecoverable things and get updated part-way through" is "don't do dangerous, irrecoverable things". 

This example is meant to highlight that OLM will not start the upgrade process while an operator is communicating that it cannot be upgraded to the next version.

>A better example for this enhancement might be "later versions of my operator, which you might be considering updating to, are not compatible with the current cluster state. If you update me to those versions, I will break". Or something like that?

As @ecordell [discussed](https://github.com/operator-framework/enhancements/pull/23#discussion_r474012292), this example is slightly different than the issues we are trying to address in this enhancement.